man/man1/logtalk_user_setup.1
.TH logtalk_user_setup 1 "September 16, 2019" "Logtalk 3.30.0" "Logtalk Documentation"

.SH NAME
logtalk_user_setup \- Logtalk user setup script

.SH SYNOPSIS
.B logtalk_user_setup

.SH DESCRIPTION
\fBlogtalk_user_setup\fR is a Logtalk user setup script for copying the Logtalk user-modifiable files and directories to the location pointed by the environment variable LOGTALKUSER (defaults to "~/logtalk"). When the Logtalk user directory already exists, a backup is made to a directory whose name starts with "logtalk-backup-" and a new Logtalk user directory is created.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
logtalk_version_select(1),\ logtalk_backend_select(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2021 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
Copyright 1998-2021 Paulo Moura <pmoura@logtalk.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.44.0 - January ??, 2021
=========================

Documentation
-------------

* IMPROVED: Documentation of the `logtalk::loaded_file_property/2` predicate.

Library
-------

* ADDED: Implementation for the `os::directory_files/2` predicate for Trealla
ProLog.

* FIXED: Implementation for the `os::delete_directory/1` predicate for Trealla
ProLog.

Installers and installation scripts
-----------------------------------

* CHANGED: Renamed the `scripts/macosx` directory to `scripts/macos`.

* ADDED: Windows installer experimental support for creating a shortcut for
Tau Prolog.

* ADDED: A macOS command-file for Trealla ProLog.

* ADDED: A `update_man_html_versions.sh` shell script for updating the HTML
versions of the man pages.

* UPDATED: The macOS installer to also set system-wide default values for the
`LOGTALKHOME` and `LOGTALKUSER` environment variables for the `zsh` shell.


3.43.0 - December 22, 2020
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Lint check for redundant use of the `call/1` control construct as
argument to the `\+/1` built-in predicate.

* ADDED: Also check correctness of `variable_names/1` options in calls to the
`write_term/2-3` Prolog standard predicates.

* IMPROVED: Compiler performance for protocols.

* IMPROVED: More efficient compiler checking for parameter variables.

* FIXED: Using a overriding meta-predicate template for predicates defined in
`user` listed in `uses/2` directives.

* FIXED: Bug in the `logtalk_make/0-1` predicates when reloading a main file
due to an updated included file when the `clean` flag is turned off and the
`reload` file is set tp `changed` where the main file would not be recompiled
to include the new contents of the included file.

* FIXED: Bug in the `asserta/1` built-in method when the argument is only know
at runtime.

Prolog adapter and integration files
------------------------------------

* ADDED: Adapter file for Trealla Prolog. Experimental. With the help of
Andrew Davison.

* IMPROVED: Adapter files definition of the `{}/1` top-level interpreter
shortcut for loading files to use a single call to the `logtalk_load/1`
built-in predicate.

* UPDATED: The LVM adapter file to take advantage of indexing support for the
internal runtime tables and of the new `atom_hash/2` built-in predicate.

* UPDATED: The LVM integration script to select the standard-top-level.

* UPDATED: The Tau Prolog and Trealla ProLog adapter files for Unicode support.

Documentation
-------------

* ADDED: Man page for the Trealla ProLog integration script, `tplgt`.

* FIXED: Mention of settings file possible locations in the integration script
man pages.

Library
-------

* ADDED: Support for Trealla ProLog to the `os` and `random` libraries.

* ADDED: Library `git` with predicates for accessing a git project current
branch and latest commit data.

* ADDED: Definition for the `os::wall_time/1` predicate for Tau Prolog.

* ADDED: New `jump_all_block/3`, `append/3`, and `map/2-3` predicates to the
`queues` library.

* ADDED: Tests for the `os` library `wall_time/1` and `shell/1-2` predicates.

* UPDATED: The `hook_objects` library `object_wrapper_hook` object to set the
`context_switching_calls` flag to `allow` for the generated objects, enabling
calling the wrapped predicates using the `<</2` context-switching control
construct.

* FIXED: The `os::directory_files/2` predicate for LVM to also return the `.`
and `..` directories as per the `osp` protocol.

* FIXED: Test failure for the `hook_objects` library due to missing setup step.

* FIXED: Fix the `os::shell/1` library predicate for ECLiPSe (broken due to a
bug in its `system/1` built-in predicate).

* FIXED: The `os::command_line_arguments/1` predicate for Tau Prolog.

* FIXED: The `os` library predicates `cpu_time/1` and `wall_time/1` for LVM.

Tools
-----

* CHANGED: The `lgtunit` tool now requires and checks that test identifiers
are ground callable terms.

* ADDED: Predicates `set_text_output/3` and `set_binary_output/3` to the
`lgtunit` tool.

* ADDED: Support for Trealla ProLog to the `debugger`, `lgtunit`, and
`ports_profiler` tools.

* ADDED: Additional test for the `debugger::trace/0` predicate.

* IMPROVED: Advice given by the `tutor` tool for warnings generated by compile
time references to unknown entities.

* IMPROVED: Documentation of the `ports_profiler` tool.

* UPDATED: The `logtalk_tester` script support for the LVM backend.

* FIXED: The `lgtunit::run_test_sets/1` predicate to type-check the test set
objects.

* FIXED: The `lgtunit` tool expansion of `discontiguous/1` directives to apply
only to entity directives.

* FIXED: The `debugger` tool to call the `reset/0` predicate at loading time
to ensure that code compiled in debug mode will also run even when not tracing
or debugging. Thanks to Camilo Correa Restrepo for the bug report.

* FIXED: The `lgtdoc` tool to abolish the temporary objects created to find
inherited predicates when documenting entities.

Examples
--------

* ADDED: Crude solution to the `bench` example to run all benchmarks.

* ADDED: Tests for some of the individual benchmarks in the `bench` example.


scripts/logtalk_user_setup.js
/////////////////////////////////////////////////////////////////////////////
//
//   Logtalk user folder setup script
//   Last updated on April 9, 2019
//
//   This file is part of Logtalk <https://logtalk.org/>  
//   Copyright 1998-2021 Paulo Moura <pmoura@logtalk.org>
//   
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.
//
/////////////////////////////////////////////////////////////////////////////



if (ScriptEngineMajorVersion() < 5 || ScriptEngineMajorVersion() == 5 && ScriptEngineMinorVersion() < 6) {
	WScript.Echo('Error! WSH 5.6 or later version needed for running this script.');
	WScript.Quit(1);
}

var WshShell = new ActiveXObject("WScript.Shell");

var WshSystemEnv = WshShell.Environment("SYSTEM");
var WshUserEnv = WshShell.Environment("USER");
var logtalk_home;
var logtalk_user;

if (WshSystemEnv.Item("LOGTALKHOME"))
	logtalk_home = WshSystemEnv.Item("LOGTALKHOME");
else if (WshUserEnv.Item("LOGTALKHOME"))
	logtalk_home = WshUserEnv.Item("LOGTALKHOME")
else {
	WScript.Echo("The environment variable LOGTALKHOME must be defined first!");
	usage_help();
	WScript.Quit(1);
}

var FSObject = new ActiveXObject("Scripting.FileSystemObject");

if (!FSObject.FolderExists(logtalk_home)) {
	WScript.Echo("The environment variable LOGTALKHOME points to a non-existing directory!");
	WScript.Echo("Its current value is: %LOGTALKHOME%");
	WScript.Echo("The variable must be set to your Logtalk installation directory!");
	WScript.Echo("");
	usage_help();
	WScript.Quit(1);
}

if (WScript.Arguments.Unnamed.Length > 0)
	usage_help();

if (WshUserEnv.Item("LOGTALKUSER"))
	logtalk_user = WshUserEnv.Item("LOGTALKUSER");
else {
	logtalk_user = WshShell.SpecialFolders("MyDocuments") + "\\logtalk";
	WshUserEnv.Item("LOGTALKUSER") = WshShell.SpecialFolders("MyDocuments") + "\\logtalk";
	WScript.Echo("Defined user environment variable LOGTALKUSER.");
	WScript.Echo("");
}

if (FSObject.FolderExists(logtalk_user)) {
	var today = new Date();
	var year  = today.getFullYear();
	var month = today.getMonth() + 1;
	if (month < 10)
		month = "0" + month;
	var day = today.getDate();
	if (day < 10)
        day = "0" + day;
	var hours = today.getHours();
	if (hours < 10)
        hours = "0" + hours;
	var mins = today.getMinutes();
	if (mins < 10)
        mins = "0" + mins;
	var secs = today.getSeconds();
	if (secs < 10)
		secs = "0" + secs;
	date = year + "-" + month + "-" + day + "-" + hours + mins + secs;
	FSObject.MoveFolder(logtalk_user, logtalk_user + "-backup-" + date);
	WScript.Echo("Created a backup of the existing \"\%LOGTALKUSER\%\" directory:");
	WScript.Echo("");
	WScript.Echo("  " + logtalk_user + "-backup-" + date);
	WScript.Echo("");
	if (FSObject.FileExists(logtalk_user + "-backup-" + date + "\\settings.lgt")) {
		FSObject.CopyFile(logtalk_user + "-backup-" + date + "\\settings.lgt", logtalk_user + "\\settings.lgt");
		WScript.Echo("Copied your old \"settings.lgt\" file to the new \"\%LOGTALKUSER\%\" directory.");
	}
	if (FSObject.FileExists(logtalk_user + "-backup-" + date + "\\settings.logtalk")) {
		FSObject.CopyFile(logtalk_user + "-backup-" + date + "\\settings.logtalk", logtalk_user + "\\settings.logtalk");
		WScript.Echo("Copied your old \"settings.logtalk\" file to the new \"\%LOGTALKUSER\%\" directory.");
	}
	WScript.Echo("If you are upgrading from a previous Logtalk version, check the file");
	WScript.Echo("\"\%LOGTALKUSER\%\\settings-sample.lgt\" for updated settings suggestions.");
	WScript.Echo("");
}

WScript.Echo("Creating a new \"\%LOGTALKUSER\%\" directory:");
WScript.Echo("");
WScript.Echo("  " + logtalk_user);
WScript.Echo("");
FSObject.CreateFolder(logtalk_user);
WScript.Echo("Copying Logtalk files and directories...");
FSObject.CopyFolder(logtalk_home + "\\contributions", logtalk_user + "\\contributions");
FSObject.CopyFolder(logtalk_home + "\\docs", logtalk_user + "\\docs");
FSObject.CopyFolder(logtalk_home + "\\examples", logtalk_user + "\\examples");
FSObject.CopyFolder(logtalk_home + "\\library", logtalk_user + "\\library");
FSObject.CopyFolder(logtalk_home + "\\ports", logtalk_user + "\\ports");
FSObject.CopyFolder(logtalk_home + "\\scratch", logtalk_user + "\\scratch");
FSObject.CopyFolder(logtalk_home + "\\tests", logtalk_user + "\\tests");
FSObject.CopyFolder(logtalk_home + "\\tools", logtalk_user + "\\tools");
FSObject.CopyFile(logtalk_home + "\\loader-sample.lgt", logtalk_user + "\\loader-sample.lgt");
FSObject.CopyFile(logtalk_home + "\\settings-sample.lgt", logtalk_user + "\\settings-sample.lgt");
FSObject.CopyFile(logtalk_home + "\\tester-sample.lgt", logtalk_user + "\\tester-sample.lgt");
FSObject.CopyFile(logtalk_home + "\\tests-sample.lgt", logtalk_user + "\\tests-sample.lgt");
FSObject.CopyFile(logtalk_home + "\\VERSION.txt", logtalk_user + "\\VERSION.txt");

FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2*.*");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.dtd");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.xsd");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.rng");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.dtd");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.xsd");
FSObject.DeleteFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.rng");

var link = WshShell.CreateShortcut(logtalk_user + "\\adapters.lnk");
link.Description = "Shortcut to Logtalk adapter files";
link.TargetPath = logtalk_home + "\\adapters";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\manuals.lnk");
link.Description = "Shortcut to Logtalk documentation";
link.TargetPath = logtalk_home + "\\manuals";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\paths.lnk");
link.Description = "Shortcut to Logtalk default library and example paths";
link.TargetPath = logtalk_home + "\\paths";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\coding.lnk");
link.Description = "Shortcut to Logtalk text editing support";
link.TargetPath = logtalk_home + "\\coding";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\diagrams\\lgt2svg.js.lnk");
link.Description = "Shortcut to lgt2svg script";
link.TargetPath = logtalk_home + "\\tools\\diagrams\\lgt2svg.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2html.js.lnk");
link.Description = "Shortcut to lgt2html script";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\lgt2html.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2pdf.js.lnk");
link.Description = "Shortcut to lgt2pdf script";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\lgt2pdf.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2xml.js.lnk");
link.Description = "Shortcut to lgt2xml script";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\lgt2xml.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2md.js.lnk");
link.Description = "Shortcut to lgt2md script";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\lgt2md.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\lgt2txt.js.lnk");
link.Description = "Shortcut to lgt2txt script";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\lgt2txt.js";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.dtd.lnk");
link.Description = "Shortcut to Logtalk entity DTD";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_entity.dtd";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.rng.lnk");
link.Description = "Shortcut to Logtalk entity RELAX NG Schema";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_entity.rng";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity.xsd.lnk");
link.Description = "Shortcut to Logtalk entity XML Schema";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_entity.xsd";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.dtd.lnk");
link.Description = "Shortcut to Logtalk index DTD";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_index.dtd";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.rng.lnk");
link.Description = "Shortcut to Logtalk index RELAX NG Schema";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_index.rng";
link.Save();

link = WshShell.CreateShortcut(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index.xsd.lnk");
link.Description = "Shortcut to Logtalk index XML Schema";
link.TargetPath = logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_index.xsd";
link.Save();

WScript.Echo("Finished copying Logtalk files and directories.");
WScript.Echo("");
WScript.Echo("You may want to customize the default Logtalk compiler flags by renaming");
WScript.Echo("and editing the \"settings-sample.lgt\" source file found in the directory");
WScript.Echo("\"\%LOGTALKUSER\%\". For more information on how to customize Logtalk and the");
WScript.Echo("working environment, consult the \"\%LOGTALKUSER\%\\CUSTOMIZE.md\" file.");
WScript.Echo("");

WScript.Quit(0);

function usage_help() {
	WScript.Echo("");
	WScript.Echo("This script copies all the Logtalk per-user files and directories to the");
	WScript.Echo("user home directory. The location can be set by the environment variable");
	WScript.Echo("\"\%LOGTALKUSER\%\" (defaults to \"MyDocuments\\logtalk\" when the variable is");
	WScript.Echo("not defined).");
	WScript.Echo("");
	WScript.Echo("Usage:");
	WScript.Echo("  " + WScript.ScriptName + " help");
	WScript.Echo("  " + WScript.ScriptName);
	WScript.Echo("");
	WScript.Quit(1);
}
